### PR TITLE
New animal management structure

### PR DESCRIPTION
**Pull request by militsasotirova**
New animal management structure. This includes Manfei's animal life cycle code, which has already been approved in a previous pull request (#22).

### New input file: animal_management.json
- _herd_information_: numbers of each class of animal to initiate herd with
- _animal_config_: configuration information needed in animal life cycle
- _pen_information_: holds information specific to each pen

### New file structure within RUFAS.routines.animal
- _life_cycle_ folder: holds all of the animal & event classes as well as the LifeCycleManager class which calls their methods
- _manure_  and _ration_ folder: these folders hold the manure and ration calculation information for each of the four animal groups (calves, growing heifers, lactating cows, and dry cows). There are placeholders for those methods which have yet to be implemented.

### How the code is set up
_animal_management.py_ replaced the previous _animal.py_. It initializes the pens, animals, and life cycle and manages these segments as the simulation progresses. For a majority of the methods in this class, tasks are delegated to separate pens, which call methods in the animal classes. 

Note that the code does not currently work as Manfei's herd simulation did (in fact, the code in this pull request currently crashes), but this is the goal. So far, mostly everything has been set up for the addition of pens, different animal classes, and Manfei's life cycle code as well as the modularity of manure and ration formulations.

### What still needs to be done
- Get same graphical output as Manfei
- Connect the ration and manure code to the simulation (currently hardcoded) and deal with an infeasible ration formulation
- Create output files for new characteristics

Any feedback would be greatly appreciated! Thank you!

### CODE REVIEW

**Comment by militsasotirova** [comment]
The code now works with the additional functionality of pens and an animal management class as described in my initial comment. Ration and manure code are called and new output files with updated characteristics are produced. The simulation works with animal_management.json, barnyard.json, input_w_feedlib.json, testing_manure.json, and testing_ration_feedlib.json.

Two notes:
1. Line 65 in simulation_engine.py can be uncommented for animal graphs and end of simulation statistics to appear when the simulation has completed. 
2. Concerning delta body weight calculation (line 208 of lactating_cow_ration.py, A.ER.5.4 in pseudocode): This value is (for now) hardcoded to 0 because sometimes it would work out that 280 - CI + DOP = 0 (CI = calving interval, DOP = days of pregnancy) for a certain cow, giving a division by 0 error in that calculation. Once Li finds a different model for delta body weight in ration, this will be updated.

File: RUFAS/routines/animal/animal_management.py
Comment (KFosterReed) [inline RUFAS/routines/animal/animal_management.py:35]: Is there a reason this class variable name is in camel case? Did we decide to use camel case for specific variables? Or is it because we have already used animal_management? If it is the latter, could we find another name for this variable?

Comment (militsasotirova) [reply RUFAS/routines/animal/animal_management.py:35]: It's the Python convention to use upper camel case for class names

**Review by KFosterReed — COMMENTED** [review]
Overall I like the modularity of it. Great work Militsa! I am going to try to run it in the next day or two but the first look is great!

**Comment by militsasotirova** [comment]
You can now view graphical representations of the manure and milk production outputs if you uncomment line 67 in simulation_engine.py (if you want to see Manfei's output, uncomment line 66).

**Review by wmaxdonovan — COMMENTED** [review]
Most of the comments Jacob and I have pertain to general use rather than the code itself. 

The first thing is that the program takes much longer to run. It doesn't seem to be a problem with creating graphics (which we've found takes up a lot of CPU power) as much as the size of the herd being simulated. My understanding is that each one of these cows is created as an object that is accessed and adjusted on a somewhat-daily basis. I like this organization and it's easy to follow, but it explains why this update adds so much simulation time. I will be interested in seeing whether our meeting re: computing efficiency yields any new strategies to optimize this process. In the meantime, we need to come up with a meaningful way to toggle animal_management before merging this into the master branch. Reducing the herd size brings down simulation time but yields unexpected errors in a number of cases/locations, most notably when the herd size is 0. Aside from duration, we'll want to be able to run the program with 0 cows for data sets that don't include them. We also had issues for herd sizes between 0 and the current values, and I don't know whether it's a matter of ratios or amounts.

Similar toggles should be implemented for creating and displaying graphics, and graphics should be saved as program output if they are created. This is something that we do for the creation of the reports with the "active" boolean in the JSON file, but updating this boolean for the manure and growth reports does not stop the program from attempting to produce graphics (which causes an error because they are expecting CSVs that are not there). Incorporating graphics into the existing report handler structure generalizes the process of creating graphs for multiple reports and allows you to toggle their creation. If we want to be producing these graphs in the long term for multiple program outputs we should consolidate that code and incorporate it as a step in the simulation in the same way we create reports.

I think the data being printed to the console at the end of the simulation should be saved to a CSV both to clean up output and to make that data more easily accessible for analysis tools that we develop now/later. The console is also a temporary interface, so data that gets saved here is going to have to migrate to a different location at some point (probably a long time from now but still).

I'm interested in the simulation_duration stuff because it's something that we considered doing when we were developing the routines for perennial crops but ultimately avoided because we index our arrays using year/jday everywhere else and wanted to keep the model's conception of time consistent. It doesn't look like you're using it for indexing, but if you want year-day/year-month for your output instead of simulation day, matplotlib supports datetime objects that can be created from the existing year/jday structure of RuFaS that you can do all sorts of fancy stuff with. I'm wary of tracking something as significant as time using different objects and methods.

This is also a ton of new input for the JSON file, and I've been thinking we should come up with ways to break down input a little bit by creating separate JSON files for different soil profiles/crops/animal structures and then referencing them in a "driver" JSON file similar to how we reference the weather CSV and the feed library. That way we could have preset data to select from for inputs that aren't currently a concern, but also the ability to make more minute adjustments. I think Jacob said something like "general and advanced settings."

Other than that, we think we could also reduce simulation time if the global variables you create were just attributes of the AnimalManagement class, but if there's a specific reason you made them global instead this is not a huge deal. I think the only global variables currently in play are the state objects, and there are good reasons to keep that as a protected class for organization and storage.

Jacob noted a few lines that exceed the 80/100/120char limits that pycharm sets for syntax and compiling, but there are no problems running the code, so maybe the inspector guy is just bluffing.

The big things were simulation length which has to do with dynamic herd sizes, and toggling/integrating output. Let us know what you think/if you have any questions.

**Comment by MBarry2100** [comment]
Pardon me for interjecting, have mostly been lurking in the shadows, but as I understand it, the diet is least cost optimized via LP for each cow on every day. I don't think this would ever be applied in practice in the real world (at least for the foreseeable future). Mostly for the same reason, computational expense, not to mention the labor/technology costs of executing the diets.

T would strongly suggest formulating diets every two weeks (at the very most) and, rather than formulating for individual cows, try to formulate for across logical groupings aggregated by age, production and/or reproductive status.

Just my two cents, for what it's worth.

**Comment by militsasotirova** [comment]
Each module now can be toggled to be simulated or not.

Input files that have been updated to work with new changes: animal_management.json, barnyard.json, input_w_feedlib.json, testing_manure.json, testing_ration_feed.json.

If a module is set to inactive but the related output csv files are active in the "output" section of the json file, the reports for those csv files are automatically set to inactive.

Note: since they are dependent on each other, the soil and crop modules must be running together (one cannot run without the other)

**Comment by KFosterReed** [comment]
> Pardon me for interjecting, have mostly been lurking in the shadows, but as I understand it, the diet is least cost optimized via LP for each cow on every day. I don't think this would ever be applied in practice in the real world (at least for the foreseeable future). Mostly for the same reason, computational expense, not to mention the labor/technology costs of executing the diets.
> 
> T would strongly suggest formulating diets every two weeks (at the very most) and, rather than formulating for individual cows, try to formulate for across logical groupings aggregated by age, production and/or reproductive status.
> 
> Just my two cents, for what it's worth.

Mike thanks for jumping in and adding your comments! 

We definitely understand that the scenario of formulating a diet for every cow on every day is not a practical scenario. However, because we want this to be a research tool that could investigate this extreme scenario, we have built the model to be able to handle this case. However, the user has the ability to define how often diets are formulated (any where from once a day to once per year) and the user will also define the number of animals that are fed each ration by defining the pen size and grouping strategies. The current logic is that animals are allocated to pens and diets are formulated for the 'average' requirements within a pen. If the user wanted to feed individual animals, they would create pens that only hold one cow. 

Would love to talk more with you about this in the future!

Kristan

**Comment by militsasotirova** [comment]
Regarding the output graphs that you can display for milk production and manure excretion, these outputs are something that can be generating with the code in the OA/WB.final branch and are currently coded as they are in this branch for presentation purposes for the annual meeting. I can remove the graph generation code from this branch after the meeting and then this branch can be merged with master. As for the final output stats, I can create a toggle in the input json file for the creation of a csv file to display them once this branch and OA/WB.final have been merged/conflicts are resolved so that I can follow the output file generation structure laid out in the OA/WB.final branch.

I'd love to hear your feedback for this plan. Thanks for you input!

**Comment by militsasotirova** [comment]
The herd_num parameter in the json file now acts as a toggle for the animal module: if herd_num = 0, the boolean simulate_animals is set to False and no animals are simulated (nothing happens in the AnimalManagement’s daily_updates() method). Additionally, a message (“herd_num is 0 -> no animals will be simulated”) is printed to the console.
If herd_num = 0 but any one of calf_num, heiferI_num, heiferII_num, heiferIII_num, or cow_num is not, then a warning message is printed (e.g. “Warning: herd_num is 0, but calf_num is not. Setting calf_num = 0.”) and the corresponding variable is set to 0.

**Comment by KFosterReed** [comment]
Thanks Militsa and Tayler for implementing and checking this update. This is exactly the type of toggle we are looking for. 

For the case where there are calves and heifers but no lactating cows, I believe that is a scenario that Manfei needs to work on within her life cycle module. It is not a big priority for the model but it is probably better to build it into the structure now than to wait until later. 

> Thank you Militsa. This change does allow you to run the other components of the model without the animals (and is super fast :). Minor typos on lines 125, 128, 131, and 133 likely from copying and pasting. Although this fixes the ability to run with no animals, I tried running with just heifers and no milking cows (as if all the heifers were sold). You run into the divide by zero error in the ration formulation program. We will need to address this issue in the future. Is there another option besides using if statements?

**Comment by wmaxdonovan** [comment]
Toggles look great and we're blown away by speed! We merged output handling prior to implementation so I'm going through now and updating animal output so that it is compatible. Would it be alright for us to report output for each pen separately? This fits the structure of multiple-field output and would allow us to produce individualized as well as summary reports.

**Comment by KFosterReed** [comment]
> Toggles look great and we're blown away by speed! We merged output handling prior to implementation so I'm going through now and updating animal output so that it is compatible. Would it be alright for us to report output for each pen separately? This fits the structure of multiple-field output and would allow us to produce individualized as well as summary reports.

I think having output by pen would be a great option. Would it also be possible to have an option summarize outputs by animal class? Not an immediate need but would be useful

**Comment by wmaxdonovan** [comment]
I integrated animal outputs into the output handling/graphic production structure Jacob and I had developed and broke output up by pen. Any outstanding merge conflicts with master relating to output handling are syntactical and resolvable by defaulting to the Head (new-animal-management-structure). Flagging this bc/ these features were updated in a separate branch and reviewed by Militsa but it looks like review by @tlh96 , @KFosterReed , and @manfei-L may still be necessary. Militsa has resolved all of our concerns and I believe this merge is good to go.

File: RUFAS/output/soil_phosphorus.py
Comment (tlh96) [inline RUFAS/output/soil_phosphorus.py]: Do @Jacob8399  and @wmdonovan-uwcs want to add your emails to the author list in this file to keep it similar to other files?

**Comment by wmaxdonovan** [comment]
Fixed output related errors (missing header for ration report and conflicting csv/graphics toggles e.g. graphics requested but no csv's produced to work from). Standardized headers for output files and added author names/emails.

**Review by tlh96 — APPROVED** [review]
Hi All,
The recent changes have decreased the amount of time it takes to run the simulation, yay! I have a few minor comments for changes. @wmdonovan-uwcs  and @Jacob8399, would you mind filling in some initial variables for the annual pen reports? The growth report could include annual growth and annual milk production; manure report, MN (manure nitrogen), VSd, and VSnd; and ration report, price, and the different feeds. @militsasotirova is the separate code for graphing in the animal module still necessary given we have updated the output to produce similar graphics for all the modules? I was able to comment out lines 491 to 555 in life_cycle.py. 

Militsa, could you take charge of merging the branch into master?

Great job everyone!

**Comment by militsasotirova** [comment]
I technically can't merge the branch yet since there's only 1 approval on it.

**Comment by tlh96** [comment]
Who were the original reviewers? I took over for @KFosterReed.

**Comment by militsasotirova** [comment]
Manfei, Max, and Jacob. You should be able to see the reviewers at the top of this page on the right.

**Review by KFosterReed — APPROVED** [review]

**Comment by KFosterReed** [comment]
I just added an approval so you should be able to merge it now. 


> Manfei, Max, and Jacob. You should be able to see the reviewers at the top of this page on the right.

**Pull request merged**
